### PR TITLE
Properly reset worfklow expiration time if worfklow was reset

### PR DESCRIPTION
**Pull request by ychebotarev**
## What changed?

1. Main change - isValidExecutionTimeoutTask will return false if workflow execution timeout was fired before it actually expire. This is done to prevent workflow from instantly timing out after reset.
See https://github.com/temporalio/temporal/issues/1913 for details
2. "Refresh" timed tasks after reset. This is done because workflow execution timeout tasks are created from start events, and thus use old time.
3. Fix tests.
4. Remove unused parameters from TaskRefresherImpl

## Why?

See https://github.com/temporalio/temporal/issues/1913 for details
In short - workflow instantly times out if reset happen after workflow execution timeouts.

## How did you test it?

Create a separate workflow that reproduce this behavior, confirms that it doesn't instantly times out now.
Fix unit tests.

## Documentation

N/A
## Is hotfix candidate?

N.A

### CODE REVIEW

File: service/history/timer_queue_task_executor_base.go [hunk 1]
```diff
@@ -185,19 +185,15 @@ func (t *timerQueueTaskExecutorBase) isValidExecutionTimeoutTask(
...
-	return true
+	// fix for https://github.com/temporalio/temporal/issues/1913
+	// workflow timeout is not expired
+	// This can happen if the workflow is reset since reset re-calculates
+	// the execution timeout but shares the same firstRunID as the base run
 
-	// TODO: uncomment the following logic when fixing
-	// https://github.com/temporalio/temporal/issues/1913
+	now := t.shardContext.GetTimeSource().Now()
+	expired := queues.IsTimeExpired(now, executionInfo.WorkflowExecutionExpirationTime.AsTime())
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/timer_queue_task_executor_base.go:193]: add a separate test for expiration time and isValidExecutionTimeoutTask

File: service/history/ndc/workflow_resetter.go [hunk 2]
```diff
@@ -283,6 +279,17 @@ func (r *workflowResetterImpl) prepareResetWorkflow(
...
 	}
 
+	// if workflow was reset after it was expired - at this point expiration task will
+	// already be fired since it is (re)created from the event, and event has old expiration time
+	// generate workflow execution task. again. this time with proper expiration time
+	taskRefresher := workflow.NewTaskRefresher(
+		r.shardContext,
+		r.logger,
+	)
+	if err := taskRefresher.RefreshTasks(ctx, resetMutableState); err != nil {
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/ndc/workflow_resetter.go]: test added, but not sure if it is enough. I think I would prefer func test, but I don't know how to properly organize it without actual timeouts.

File: service/history/workflow/task_refresher.go [hunk 3]
```diff
@@ -199,6 +191,15 @@ func (r *TaskRefresherImpl) refreshTasksForWorkflowStart(
 	}
 
 	return nil
+
+}
+
+func (r *TaskRefresherImpl) refreshTasksForWorkflowStart(
```
Comment (prathyushpv) [inline service/history/workflow/task_refresher.go:197]: Why do we need this function?

File: service/history/ndc/workflow_resetter_test.go [hunk 4]
```diff
@@ -972,3 +970,106 @@ func (s *workflowResetterSuite) TestPagination() {
...
+		baseBranchToken,
+		baseRebuildLastEventID,
+		baseRebuildLastEventVersion,
+		s.resetRunID,
+		resetRequestID,
+		resetWorkflowVersion,
+		resetReason,
+	)
+	s.NoError(err)
+	s.Equal(resetMutableState, resetWorkflow.GetMutableState())
```
Comment (prathyushpv) [inline service/history/ndc/workflow_resetter_test.go:1074]: We will definitely need a test to make sure reset workflow is timed out properly. That is a timeout task is generated and it is fired as expected.

File: service/history/workflow/task_refresher.go (hunk 3 above)
Comment (ychebotarev) [reply service/history/workflow/task_refresher.go:197]: not needed anymore, remove in the next rev

File: service/history/ndc/workflow_resetter_test.go (hunk 4 above)
Comment (ychebotarev) [reply service/history/ndc/workflow_resetter_test.go:1064]: This test is checking that timeout was actually generated.
I don't think that we need a test that waits that timeout is actually fired.
Also all of that is kind of not related - reason why workflow was instantly timing out is fixed and covered by unit test.

Comment (prathyushpv) [reply service/history/ndc/workflow_resetter_test.go:1064]: But the unit test is not covering RefreshTasksForWorkflowStart() right?

File: service/history/ndc/workflow_resetter_test.go [hunk 5]
```diff
@@ -283,22 +281,15 @@ func (s *workflowResetterSuite) TestReplayResetWorkflow() {
...
 	baseRebuildLastEventVersion := int64(12)
-	baseNodeID := baseRebuildLastEventID + 1
 
 	resetBranchToken := []byte("some random reset branch token")
 	resetRequestID := uuid.New()
 	resetHistorySize := int64(4411)
 	resetMutableState := workflow.NewMockMutableState(s.controller)
 
-	shardID := s.mockShard.GetShardID()
-	s.mockExecutionMgr.EXPECT().ForkHistoryBranch(gomock.Any(), &persistence.ForkHistoryBranchRequest{
```
Comment (prathyushpv) [inline service/history/ndc/workflow_resetter_test.go:294]: It is useful to see if the argument for ForkHistoryBranch() is what we expect. Could you change only the field in the request that was different? Sorry If I missed this in last review.

File: service/history/ndc/state_rebuilder.go [hunk 6]
```diff
@@ -198,6 +197,7 @@ func (r *StateRebuilderImpl) Rebuild(
 	// TODO: ideally the executionTimeoutTimerTaskStatus field should be carried over
 	// from the base run. However, RefreshTasks always resets that field and
 	// force regenerates the execution timeout timer task.
+	r.logger.Info("QQQQQQ RefreshTasks")
```
Comment (prathyushpv) [inline service/history/ndc/state_rebuilder.go:200]: Forgot to remove?

File: service/history/ndc/state_rebuilder.go [hunk 7]
```diff
@@ -186,6 +184,7 @@ func (r *StateRebuilderImpl) Rebuild(
 		}
 	}
 
+	r.logger.Info("QQQQQQ before CloseTransactionAsSnapshot")
```
Comment (prathyushpv) [inline service/history/ndc/state_rebuilder.go:187]: Forgot to remove?

File: service/history/workflow/task_refresher.go [hunk 8]
```diff
@@ -162,7 +154,7 @@ func (r *TaskRefresherImpl) RefreshTasks(
 	)
 }
 
-func (r *TaskRefresherImpl) refreshTasksForWorkflowStart(
```
Comment (prathyushpv) [inline service/history/workflow/task_refresher.go:165]: It still makes sense for this function to be part of the TaskRefresher interface. If this function doesn't need to be a part of TaskRefresher, the same applies to RefreshTasks() also.

File: service/history/timer_queue_active_task_executor_test.go [hunk 9]
```diff
@@ -1727,6 +1727,9 @@ func (s *timerQueueActiveTaskExecutorSuite) TestWorkflowExecutionTimeout_Fire()
 	s.mockExecutionMgr.EXPECT().GetWorkflowExecution(gomock.Any(), gomock.Any()).Return(&persistence.GetWorkflowExecutionResponse{State: persistenceMutableState}, nil)
 	s.mockExecutionMgr.EXPECT().UpdateWorkflowExecution(gomock.Any(), gomock.Any()).Return(tests.UpdateWorkflowExecutionResponse, nil)
 
+	// advance the clock to be sure workflow is expired
```
Comment (prathyushpv) [inline service/history/timer_queue_active_task_executor_test.go:1730]: Curious why this was not needed before?

**Review by prathyushpv — COMMENTED** [review]
it is better to have a functional test to cover this scenario.

File: service/history/ndc/workflow_resetter_test.go (hunk 4 above)
Comment (ychebotarev) [reply service/history/ndc/workflow_resetter_test.go:1064]: unit test is covering the code that was causing the problem, which is not RefreshTasksForWorkflowStart.
this test is making sure RefreshTasksForWorkflowStart is called.

File: service/history/ndc/state_rebuilder.go (hunk 6 above)
Comment (ychebotarev) [reply service/history/ndc/state_rebuilder.go:200]: yes. removed in the next rev

File: service/history/ndc/state_rebuilder.go (hunk 7 above)
Comment (ychebotarev) [reply service/history/ndc/state_rebuilder.go:187]: yes, removed in the next rev

File: service/history/workflow/task_refresher.go (hunk 8 above)
Comment (ychebotarev) [reply service/history/workflow/task_refresher.go:165]: can you explain why it should be part of TaskRefresher interface?

File: service/history/timer_queue_active_task_executor_test.go (hunk 9 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor_test.go:1730]: that was a part of a bug - timeout task triggered before it should be, but still terminates the workflow.

File: service/history/workflow/task_refresher.go (hunk 8 above)
Comment (prathyushpv) [reply service/history/workflow/task_refresher.go:165]: I don't see any reason for that. The only reason I see is, it should be grouped with RefreshTasks().
In that case, it looks like it is better to get rid of TaskRefresher altogether.

Comment (ychebotarev) [reply service/history/workflow/task_refresher.go:165]: Here is a long explanation.
RefreshTasksForWorkflowStart needs taskGenerator.  Mutable state has taskGenerator.

But TaskRefresherImpl:
1. Create new taskGenerator (will it be the same? probably, but who knows)
2. Need shardContext, that is not needed for RefreshTasksForWorkflowStart

So options are :

1. 
   a. add RefreshTasksForWorkflowStart to TaskRefresher interface, then introduce subtle side-effect (it is using different generator. may work for now, but this is still side-effect, and just clumsy). 
   b. find a way to "pass through" shard context and logger to that call, even if both those parameters are not used in RefreshTasksForWorkflowStart.

2. Don't do that, just define RefreshTasksForWorkflowStart function that will accept only needed arguments.

Option 1. is very clumsy. So I choose 2.

File: service/history/ndc/workflow_resetter_test.go (hunk 5 above)
Comment (prathyushpv) [reply service/history/ndc/workflow_resetter_test.go:294]: As we discussed over slack, I agree with your reasoning that ForkHistoryBranch() is not called directly by replayResetWorkflow() and this check may be unnecessary.

**Review by prathyushpv — APPROVED** [review]
